### PR TITLE
[TTI][RISCV] Implement cost of some intrinsics with LMUL

### PR DESCRIPTION
**Pull request by LiqinWeng**
Intrinsics include: sadd_sat/ssub_sat/uadd_sat/usub_sat/fabs/fsqrt/cttz/ctlz/ctpop

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 1]
```diff
@@ -1013,20 +1013,78 @@ RISCVTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
...
+               getRISCVInstructionCost(RISCV::VFSGNJX_VV, LT.second, CostKind);
+    }
+    break;
+  }
   case Intrinsic::sqrt: {
     auto LT = getTypeLegalizationCost(RetTy);
-    if (ST->hasVInstructions() && LT.second.isVector())
-      return LT.first;
+    // FIXME: not get the correct cost about the llvm.sqrt.vxbf16
+    // LT.second is promote llvm::MVT::f32
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1062]: Not for this PR, but we probably want to handle the zvfhmin/zvfbfmin promotion for all intrinsics. We could pull this out of the switch statement later.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 2]
```diff
@@ -1013,20 +1013,78 @@ RISCVTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
...
+        break;
+      }
+      return LT.first * getRISCVInstructionCost(Op, LT.second, CostKind);
+    }
+    break;
+  }
+  case Intrinsic::fabs: {
+    auto LT = getTypeLegalizationCost(RetTy);
+    // FIXME: not get the correct cost about the llvm.sqrt.vxbf16
+    // LT.second is promote llvm::MVT::f32
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1041]: Is this comment still needed? The costs in the tests look correct

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 3]
```diff
@@ -1013,20 +1013,78 @@ RISCVTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
...
+    if (ST->hasVInstructions() && LT.second.isVector()) {
+      SmallVector<unsigned, 3> Opcodes;
+      if (LT.second.getVectorElementType() == MVT::bf16)
+        Opcodes = {RISCV::VFWCVTBF16_F_F_V, RISCV::VFSQRT_V,
+                   RISCV::VFNCVTBF16_F_F_W};
+      else if (LT.second.getVectorElementType() == MVT::f16 &&
+               !ST->hasVInstructionsF16())
+        Opcodes = {RISCV::VFWCVT_F_F_V, RISCV::VFSQRT_V, RISCV::VFNCVT_F_F_W};
+      else
+        Opcodes = {RISCV::VFSQRT_V};
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1072]: Opcodes isn't used?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 2 above)
Comment (LiqinWeng) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: removed

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 3 above)
Comment (LiqinWeng) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: fixed

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 1 above)
Comment (LiqinWeng) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: I will create new pr deal with f16 && bf16

File: llvm/test/Analysis/CostModel/RISCV/fp-min-max-abs.ll [hunk 4]
```diff
@@ -1,28 +1,28 @@
 ; NOTE: Assertions have been autogenerated by utils/update_analyze_test_checks.py
 ; RUN: opt < %s -passes="print<cost-model>" 2>&1 -disable-output -S -mtriple=riscv64 -mattr=+v,+f,+d,+zfh,+zvfh | FileCheck %s --check-prefixes=CHECK,ZVFH
-; RUN: opt < %s -passes="print<cost-model>" 2>&1 -disable-output -S -mtriple=riscv64 -mattr=+v,+f,+d,+zfh,+zvfhmin | FileCheck %s --check-prefixes=CHECK,ZVFHMIN
+; RUN: opt < %s -passes="print<cost-model>" 2>&1 -disable-output -S -mtriple=riscv64 -mattr=+v,+f,+d,+zfh,+zvfhmin,+zvfhbmin | FileCheck %s --check-prefixes=CHECK,ZVFHMIN
```
Comment (lukel97) [inline llvm/test/Analysis/CostModel/RISCV/fp-min-max-abs.ll:3]: Should this be +zvfbfmin?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 5]
```diff
@@ -1013,20 +1013,65 @@ RISCVTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
...
+      return LT.first * getRISCVInstructionCost(Op, LT.second, CostKind);
+    }
+    break;
+  }
   case Intrinsic::fabs:
   case Intrinsic::sqrt: {
     auto LT = getTypeLegalizationCost(RetTy);
-    if (ST->hasVInstructions() && LT.second.isVector())
-      return LT.first;
+    // TODO: add f16/bf16, bf16 with zvfhbmin && f16 with zvfhmin
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1041]:
```suggestion
    // TODO: add f16/bf16, bf16 with zvfbfmin && f16 with zvfhmin
```

Comment (LiqinWeng) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: sorry , fixed

File: llvm/test/Analysis/CostModel/RISCV/fp-min-max-abs.ll (hunk 4 above)
Comment (LiqinWeng) [reply llvm/test/Analysis/CostModel/RISCV/fp-min-max-abs.ll]: fixed

File: llvm/test/Analysis/CostModel/RISCV/fp-min-max-abs.ll [hunk 6]
```diff
@@ -1,28 +1,28 @@
 ; NOTE: Assertions have been autogenerated by utils/update_analyze_test_checks.py
 ; RUN: opt < %s -passes="print<cost-model>" 2>&1 -disable-output -S -mtriple=riscv64 -mattr=+v,+f,+d,+zfh,+zvfh | FileCheck %s --check-prefixes=CHECK,ZVFH
```
Comment (lukel97) [inline llvm/test/Analysis/CostModel/RISCV/fp-min-max-abs.ll:2]: Can you add zvfbfmin to the ZVFH run line too? I think that matches the other tests with bfloat

Comment (LiqinWeng) [reply llvm/test/Analysis/CostModel/RISCV/fp-min-max-abs.ll]: done

File: llvm/test/Analysis/CostModel/RISCV/fp-min-max-abs.ll [hunk 7]
```diff
@@ -50,21 +50,37 @@ define void @fabs() {
...
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %15 = call <vscale x 4 x half> @llvm.fabs.nxv4f16(<vscale x 4 x half> undef)
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %16 = call <vscale x 8 x half> @llvm.fabs.nxv8f16(<vscale x 8 x half> undef)
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 4 for instruction: %17 = call <vscale x 16 x half> @llvm.fabs.nxv16f16(<vscale x 16 x half> undef)
 ; CHECK-NEXT:  Cost Model: Found an estimated cost of 0 for instruction: ret void
 ;
   call half @llvm.fabs.f16(half undef)
+  call <2 x bfloat> @llvm.fabs.v2f16(<2 x bfloat> undef)
+  call <4 x bfloat> @llvm.fabs.v4f16(<4 x bfloat> undef)
+  call <8 x bfloat> @llvm.fabs.v8f16(<8 x bfloat> undef)
+  call <16 x bfloat> @llvm.fabs.v16f16(<16 x bfloat> undef)
```
Comment (lukel97) [inline llvm/test/Analysis/CostModel/RISCV/fp-min-max-abs.ll:75]: This should be `@llvm.fabs.v2bf16`?

File: llvm/test/Analysis/CostModel/RISCV/fp-min-max-abs.ll [hunk 8]
```diff
@@ -50,21 +50,37 @@ define void @fabs() {
...
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %12 = call <vscale x 8 x bfloat> @llvm.fabs.nxv8bf16(<vscale x 8 x bfloat> undef)
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 4 for instruction: %13 = call <vscale x 16 x bfloat> @llvm.fabs.nxv16bf16(<vscale x 16 x bfloat> undef)
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %14 = call <vscale x 2 x half> @llvm.fabs.nxv2f16(<vscale x 2 x half> undef)
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 1 for instruction: %15 = call <vscale x 4 x half> @llvm.fabs.nxv4f16(<vscale x 4 x half> undef)
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %16 = call <vscale x 8 x half> @llvm.fabs.nxv8f16(<vscale x 8 x half> undef)
+; CHECK-NEXT:  Cost Model: Found an estimated cost of 4 for instruction: %17 = call <vscale x 16 x half> @llvm.fabs.nxv16f16(<vscale x 16 x half> undef)
 ; CHECK-NEXT:  Cost Model: Found an estimated cost of 0 for instruction: ret void
 ;
   call half @llvm.fabs.f16(half undef)
+  call <2 x bfloat> @llvm.fabs.v2f16(<2 x bfloat> undef)
```
Comment (lukel97) [inline llvm/test/Analysis/CostModel/RISCV/fp-min-max-abs.ll:72]: Can you move the bf16 lines into the main @fabs() test? The f16 tests were split out just to show the difference between zvfh and zvfhmin IIRC, which bf16 won't need

File: llvm/test/Analysis/CostModel/RISCV/fp-min-max-abs.ll (hunk 7 above)
Comment (LiqinWeng) [reply llvm/test/Analysis/CostModel/RISCV/fp-min-max-abs.ll]: sorry ， fixed

File: llvm/test/Analysis/CostModel/RISCV/fp-min-max-abs.ll (hunk 8 above)
Comment (LiqinWeng) [reply llvm/test/Analysis/CostModel/RISCV/fp-min-max-abs.ll]: done

**Review by lukel97 — APPROVED** [review]
LGTM, thanks

**Pull request merged**